src/api/artifact/controller.go

```diff
@@ -407,30 +429,46 @@ func (c *controller) assembleTag(ctx context.Context, tag *tm.Tag, option *TagOp
 		return t
 	}
 	if option.WithImmutableStatus {
-		repo, err := c.repoMgr.Get(ctx, tag.RepositoryID)
-		if err != nil {
-			log.Error(err)
-		} else {
-			t.Immutable = c.isImmutable(repo.ProjectID, repo.Name, tag.Name)
-		}
+		c.populateImmutableStatus(ctx, t)
 	}
-	// TODO populate signature on tag level?
 	return t
 }
 
-// check whether the tag is Immutable
-func (c *controller) isImmutable(projectID int64, repo string, tag string) bool {
-	_, repoName := utils.ParseRepository(repo)
-	matched, err := c.immutableMtr.Match(projectID, art.Candidate{
+func (c *controller) populateLabels(ctx context.Context, art *Artifact) {
+	labels, err := c.labelMgr.ListByArtifact(ctx, art.ID)
+	if err != nil {
+		log.Errorf("failed to list labels of artifact %d: %v", art.ID, err)
+		return
+	}
+	art.Labels = labels
+}
+
+func (c *controller) populateImmutableStatus(ctx context.Context, tag *Tag) {
+	repo, err := c.repoMgr.Get(ctx, tag.RepositoryID)
+	if err != nil {
+		log.Error(err)
+		return
+	}
+	_, repoName := utils.ParseRepository(repo.Name)
+	matched, err := c.immutableMtr.Match(repo.ProjectID, art.Candidate{
 		Repository:  repoName,
-		Tag:         tag,
-		NamespaceID: projectID,
+		Tag:         tag.Name,
+		NamespaceID: repo.ProjectID,
 	})
 	if err != nil {
 		log.Error(err)
-		return false
+		return
 	}
-	return matched
+	tag.Immutable = matched
+}
+
+func (c *controller) populateScanOverview(ctx context.Context, art *Artifact) {
```

Comment:
Is it possible to define a standard populate interface with kinds of implementations for such cases?

```
func (c *controller) populateScanOverview(ctx context.Context, art *Artifact) {
	// TODO implement
}

func (c *controller) populateSignature(ctx context.Context, art *Artifact) {
	// TODO implement
	// TODO populate signature on artifact or tag level?
}	}
```

Reply:
Issue https://github.com/goharbor/harbor/issues/10741 opened to track the work